Gateway 2.8 > 3.4 LTS upgrade guide (#963)
Kong/developer.konghq.com: app/gateway/upgrade/lts-upgrade.md

* **Upgrade**: The overall process of switching from an older to a newer version of {{site.base_gateway}}. 
* **Migration**: The migration of your data store data into a new environment. 
For example, the process of moving 2.8.x data from an old PostgreSQL instance to a new one for 3.4.x is referred to as database migration.

To make sure your upgrade is successful, carefully review all the steps in this guide. It’s very important to understand all the preparation steps and choose the recommended upgrade path based on your deployment type.

{:.warning}
> **Caution**: The migration pattern described in this document can only happen between two LTS versions, {{site.ee_product_name}} 2.8 LTS and {{site.ee_product_name}} 3.4 LTS. If you apply this document to other release intervals, database modifications may be run in the wrong sequence and leave the database schema in a broken state. To migrate between other versions, see the [general upgrade guide](/gateway/upgrade/).

## Prerequisites

Read this document thoroughly to successfully complete the upgrade process, as it includes all the necessary operational knowledge for the upgrade.

> * Starting from 3.4, Cassandra is not supported. 
> If you're using Cassandra as your data store, migrate off of Cassandra first and upgrade second.

Reviewer: ```suggestion
```

> Work with the Kong support team to migrate from Cassandra to PostgreSQL.

Reviewer: ```suggestion

```

* Review version compatibility between your platform version and the version of {{site.kong_gateway}} that you are upgrading to:
  * [OS version](/gateway/version-support-policy/#supported-versions)
  * [Kubernetes version and Helm prerequisites](/kic/version-support-policy/)
  * [Hardware resources](/gateway/performance/resource-sizing-guidelines/)
  * [Database and dependency versions](/gateway/third-party-support/)
* If you're using decK, [upgrade it to the latest version](/deck/).

> 

Reviewer: ```suggestion
{:.warning}
> Starting from {{site.base_gateway}} 3.4, Cassandra is not supported. 
>If you're using Cassandra as your data store, migrate off of Cassandra first and upgrade second.
>Work with the Kong support team to migrate from Cassandra to PostgreSQL.
```

Reviewer: I felt like the original way you had it, it didn't feel like prerequisite but rather something that should be called out individually. Migrating to a different database felt different than upgrading deck.

## Upgrade journey overview

### Preparation phase

There are a number of steps you must complete before upgrading to {{site.base_gateway}} 3.4 LTS:

1. Work through any listed prerequisites.
1. [Back up](#preparation-choose-a-backup-strategy) your database or your declarative configuration files.
1. Choose the right [strategy for upgrading](#preparation-choose-an-upgrade-strategy-based-on-deployment-mode) based on your deployment topology.
1. Review the [{{site.base_gateway}} changes from 2.8 to 3.4](#preparation-review-gateway-changes) for any breaking changes that affect your deployments.
1. Conduct a thorough examination of the [modifications made to the `kong.conf` file](#kong-conf-changes) between the LTS releases.
1. Using your chosen strategy, test migration in a pre-production environment.

### Performing the upgrade

The actual execution of the upgrade depends on type of deployment you have with {{site.base_gateway}}. 
In this part of the upgrade journey, you will use the strategy you determined during the preparation phase.

1. Execute your chosen upgrade strategy on dev.
2. Move from dev to prod.
3. Smoke test.

> 4. Wrap up the upgrade or roll back and try again.

Reviewer: ```suggestion
4. Finish the roll back and try again.
```

Choose your deployment mode:
* [Traditional](#traditional-mode)
* [DB-less](#db-less-mode)
* [Hybrid](#hybrid-mode)

Here's a flowchart that breaks down how the decision process works:

{% include_cached /upgrade/flow.md %}

See the following sections for breakdowns of each strategy.

### Traditional mode

{% include_cached /upgrade/traditional.md %}

#### Dual-cluster upgrade

Upgrading {{site.base_gateway}} from one LTS version to another LTS version with zero downtime can be achieved through a [dual-cluster upgrade strategy](/gateway/upgrade/dual-cluster/). 
This approach involves setting up a new cluster running the upgraded version of {{site.base_gateway}} alongside the existing cluster running the current version.

At a high level, the process typically involves the following steps:

1. **Provisioning a same-size deployment**: You need to ensure that the new cluster, which will run the upgraded version of {{site.base_gateway}}, has the same capacity and resources as the existing cluster. This ensures that both clusters can handle the same amount of traffic and workload.

> 2. **Setting up dual-cluster deployment**: Once the new cluster is provisioned, you can start deploying your APIs and configurations to both clusters simultaneously. The dual cluster deployment allows both the old and new clusters to coexist and process requests in parallel.

Reviewer: ```suggestion
1. **Setting up dual-cluster deployment**: Once the new cluster is provisioned, you can start deploying your APIs and configurations to both clusters simultaneously. The dual cluster deployment allows both the old and new clusters to coexist and process requests in parallel.
```

3. **Data synchronization**:  During the dual cluster deployment, data synchronization is crucial to ensure that both clusters have the same data. This can involve migrating data from the old cluster to the new one or setting up a shared data storage solution to keep both clusters in sync. Import the database from the old cluster to the new cluster by using a snapshot or `pg_restore`.

> 4. **Traffic rerouting**: As the new cluster is running alongside the old one, you can start gradually routing incoming traffic to the new cluster. This process can be done gradually or through a controlled switchover mechanism to minimize any impact on users. This can be achieved by any load balancer, like DNS, Nginx, F5, or even a {{site.base_gateway}} node with Canary plugin enabled.

Reviewer: ```suggestion
1. **Traffic rerouting**: As the new cluster is running alongside the old one, you can start gradually routing incoming traffic to the new cluster. This process can be done gradually or through a controlled switchover mechanism to minimize any impact on users. This can be achieved by any load balancer, like DNS, Nginx, F5, or even a {{site.base_gateway}} node with Canary plugin enabled.
```
